File: app/utils/chat-content.ts
// Tokenize a (decrypted) chat message into renderable pieces so ChatMessageContent
// can draw mentions, links and inline images as real elements instead of one
// v-html blob. Nothing here trusts the input as HTML: text is emitted verbatim for
// Vue to escape on render, and only http(s) URLs ever become a link/image src.
//
// Mentions are stored in the plaintext as `@<userId>` (the id, not the name, so a
// rename re-renders correctly); the renderer looks the current name up by id.
export type ChatToken =
  | { type: 'text'; value: string }
  | { type: 'link'; href: string; label: string }
  | { type: 'image'; href: string }
  | { type: 'mention'; userId: string }

// A mention token `@<id>` (id: anything but whitespace, angle brackets or @), or
// an http(s) URL run.
const TOKEN_RE = /(@<[^\s<>@]+>)|(\bhttps?:\/\/[^\s<]+)/gi
// Trailing prose punctuation pulled back out of a URL.
const TRAILING = /[.,!?;:'")\]}]+$/
const IMAGE_EXT = /\.(png|jpe?g|gif|webp|avif|bmp|svg)$/i

function isImageUrl(href: string): boolean {
  try {
    return IMAGE_EXT.test(new URL(href).pathname)
  } catch {
    return false
  }
}

export function parseChatContent(text: string): ChatToken[] {
  const tokens: ChatToken[] = []
  let last = 0
  for (const m of text.matchAll(TOKEN_RE)) {
    const idx = m.index ?? 0
    if (idx > last) tokens.push({ type: 'text', value: text.slice(last, idx) })
    if (m[1]) {
      tokens.push({ type: 'mention', userId: m[1].slice(2, -1) })
      last = idx + m[1].length
    } else {
      const raw = m[2]!
      const href = raw.replace(TRAILING, '')
      tokens.push(isImageUrl(href) ? { type: 'image', href } : { type: 'link', href, label: href })
      const trailing = raw.slice(href.length)
      if (trailing) tokens.push({ type: 'text', value: trailing })
      last = idx + raw.length
    }
  }
  if (last < text.length) tokens.push({ type: 'text', value: text.slice(last) })
  return tokens
}

// The first plain (non-image) link in a message, which is the one we unfurl into
// a preview card. Null if the message has no linkable URL.
export function firstPreviewLink(tokens: ChatToken[]): string | null {
  for (const tok of tokens) if (tok.type === 'link') return tok.href
  return null
}

// The user ids @-mentioned in a message - used to flag a message as mentioning the
// reader (the unread-mention badge) without re-parsing on the render path.
export function extractMentions(text: string): string[] {
  const ids = new Set<string>()
  for (const m of text.matchAll(/@<([^\s<>@]+)>/g)) ids.add(m[1]!)
  return [...ids]
}

export function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
}

// Composer text is written/displayed with @DisplayName mentions; map each one back
// to a stable @<id> token (rename-proof) for the wire/stored form. Longest names
// first so "@John Doe" wins over "@John". The inverse of decodeMentions.
export function encodeMentions(text: string, members: readonly { userId: string; name: string }[]): string {
  let out = text
  for (const m of [...members].sort((a, b) => b.name.length - a.name.length)) {
    out = out.replace(new RegExp(`(^|\\s)@${escapeRegExp(m.name)}(?=\\s|$|[^\\w])`, 'g'), `$1@<${m.userId}>`)
  }
  return out
}

// Put a stored message back into the edit box as @DisplayName (and used by search
// so a query matches the visible name). Unknown ids fall back to unknownLabel.
export function decodeMentions(text: string, names: Record<string, string>, unknownLabel: string): string {
  return text.replace(/@<([^\s<>@]+)>/g, (_, id) => `@${names[id] ?? unknownLabel}`)
}

File: apps/mobile-flutter/parity/lib/chat_content.dart
/// The chat message wire format, ported from `apps/web-nuxt/app/utils/chat-content.ts`
/// and pinned by `shared/parity-json/chat-content.json`.
///
/// Mentions are stored in the plaintext as `@<userId>` - the id, not the name,
/// so a rename re-renders correctly - and the renderer looks the current name up
/// by id. A client that skips that step shows a raw user id where a name belongs.
library;

sealed class ChatToken {
  const ChatToken();
  Map<String, Object?> toJson();
}

class TextToken extends ChatToken {
  const TextToken(this.value);
  final String value;
  @override
  Map<String, Object?> toJson() => {'type': 'text', 'value': value};
}

class LinkToken extends ChatToken {
  const LinkToken(this.href, this.label);
  final String href;
  final String label;
  @override
  Map<String, Object?> toJson() => {'type': 'link', 'href': href, 'label': label};
}

class ImageToken extends ChatToken {
  const ImageToken(this.href);
  final String href;
  @override
  Map<String, Object?> toJson() => {'type': 'image', 'href': href};
}

class MentionToken extends ChatToken {
  const MentionToken(this.userId);
  final String userId;
  @override
  Map<String, Object?> toJson() => {'type': 'mention', 'userId': userId};
}

// A mention token `@<id>` (id: anything but whitespace, angle brackets or @), or
// an http(s) URL run.
final _tokenRe = RegExp(r'(@<[^\s<>@]+>)|(\bhttps?://[^\s<]+)', caseSensitive: false);
// Trailing prose punctuation pulled back out of a URL.
final _trailing = RegExp(r'''[.,!?;:'")\]}]+$''');
final _imageExt = RegExp(r'\.(png|jpe?g|gif|webp|avif|bmp|svg)$', caseSensitive: false);
final _mentionRe = RegExp(r'@<([^\s<>@]+)>');

bool _isImageUrl(String href) {
  final uri = Uri.tryParse(href);
  if (uri == null) return false;
  return _imageExt.hasMatch(uri.path);
}

/// Tokenize a decrypted message into renderable pieces. Nothing here trusts the
/// input as markup: text comes back verbatim for the widget layer to render.
List<ChatToken> parseChatContent(String text) {
  final tokens = <ChatToken>[];
  var last = 0;
  for (final m in _tokenRe.allMatches(text)) {
    if (m.start > last) tokens.add(TextToken(text.substring(last, m.start)));
    final mention = m.group(1);
    if (mention != null) {
      tokens.add(MentionToken(mention.substring(2, mention.length - 1)));
      last = m.start + mention.length;
    } else {
      final raw = m.group(2)!;
      final href = raw.replaceFirst(_trailing, '');
      tokens.add(_isImageUrl(href) ? ImageToken(href) : LinkToken(href, href));
      final trailing = raw.substring(href.length);
      if (trailing.isNotEmpty) tokens.add(TextToken(trailing));
      last = m.start + raw.length;
    }
  }
  if (last < text.length) tokens.add(TextToken(text.substring(last)));
  return tokens;
}

/// The user ids @-mentioned in a message, in first-seen order and deduplicated.
List<String> extractMentions(String text) {
  final ids = <String>{};
  for (final m in _mentionRe.allMatches(text)) {
    ids.add(m.group(1)!);
  }
  return ids.toList();
}

String _escapeRegExp(String s) => s.replaceAllMapped(RegExp(r'[.*+?^${}()|[\]\\]'), (m) => '\\${m[0]}');

/// Composer text written as `@DisplayName` mapped onto the stable `@<id>` form.
/// Longest names first so "@John Doe" wins over "@John".
///
/// A member with an EMPTY name is skipped, and not as tidiness: the pattern
/// would collapse to `@` followed by any non-word character, which matches the
/// `@` of a token an earlier iteration just wrote - turning every mention in the
/// message into `@<empty-id><real-id>`. One nameless member corrupted the room.
String encodeMentions(String text, List<({String userId, String name})> members) {
  var out = text;
  final ordered = [...members.where((m) => m.name.isNotEmpty)]
    ..sort((a, b) => b.name.length.compareTo(a.name.length));
  for (final m in ordered) {
    out = out.replaceAllMapped(
      RegExp('(^|\\s)@${_escapeRegExp(m.name)}(?=\\s|\$|[^\\w])'),
      (match) => '${match[1]}@<${m.userId}>',
    );
  }
  return out;
}

/// Put a stored message back into `@DisplayName` form (the edit box, and any
/// place a plain string is wanted). Unknown ids fall back to [unknownLabel].
String decodeMentions(String text, Map<String, String> names, String unknownLabel) =>
    text.replaceAllMapped(_mentionRe, (m) => '@${names[m.group(1)!] ?? unknownLabel}');
